src/util.rs

```diff
@@ -37,30 +37,33 @@ pub(crate) mod ptr {
     /// `Ptr<'a, T>` is [covariant] in `'a` and `T`.
     ///
     /// [covariant]: https://doc.rust-lang.org/reference/subtyping.html
-    pub struct Ptr<'a, T: 'a + ?Sized> {
-        // INVARIANTS:
-        // 1. `ptr` is derived from some valid Rust allocation, `A`
-        // 2. `ptr` has the same provenance as `A`
-        // 3. `ptr` addresses a byte range which is entirely contained in `A`
-        // 4. `ptr` addresses a byte range whose length fits in an `isize`
-        // 5. `ptr` addresses a byte range which does not wrap around the address
-        //     space
-        // 6. `ptr` is validly-aligned for `T`
-        // 7. `A` is guaranteed to live for at least `'a`
-        // 8. `T: 'a`
+    pub struct Ptr<'a, T: 'a + ?Sized, const ASSUME_ALIGNED: bool> {
+        /// INVARIANTS:
+        /// 1. `ptr` is derived from some valid Rust allocation, `A`
+        /// 2. `ptr` has the same provenance as `A`
+        /// 3. `ptr` addresses a byte range which is entirely contained in `A`
+        /// 4. `ptr` addresses a byte range whose length fits in an `isize`
+        /// 5. `ptr` addresses a byte range which does not wrap around the address
+        ///     space
+        /// 6. If `ASSUME_ALIGNED`, `ptr` is validly-aligned for `T`
+        /// 7. `A` is guaranteed to live for at least `'a`
+        /// 8. `T: 'a`
         ptr: NonNull<T>,
         _lifetime: PhantomData<&'a ()>,
     }
 
-    impl<'a, T: ?Sized> Copy for Ptr<'a, T> {}
-    impl<'a, T: ?Sized> Clone for Ptr<'a, T> {
+    /// The argument to `TryFromBytes::is_bit_valid`.
+    pub type MaybeValid<'a, T> = Ptr<'a, T, false>;
```

Comment:
IMO the name `MaybeValid` implies that the only thing that's not guaranteed is validity, but here lifetime and aliasing are both not guaranteed. I think we should just use `Ptr` directly since it's more clear that the user should go look at the type rather than guess its semantics.

Reply:
`Ptr` does guarantee lifetime (item 8):
```rust
/// INVARIANTS:
/// 1. `ptr` is derived from some valid Rust allocation, `A`
/// 2. `ptr` has the same provenance as `A`
/// 3. `ptr` addresses a byte range which is entirely contained in `A`
/// 4. `ptr` addresses a byte range whose length fits in an `isize`
/// 5. `ptr` addresses a byte range which does not wrap around the address
///     space
/// 6. If `ASSUME_ALIGNED`, `ptr` is validly-aligned for `T`
/// 7. `A` is guaranteed to live for at least `'a`
/// 8. `T: 'a`
```
...but not aliasing. However, this type *only* appears as an argument to `is_bit_valid` (and is documented as such), and `is_bit_valid`, as a pre-condition, guarantees the appropriate aliasing requirements. The name `MaybeValid`, then, does convey the salient quality of the argument.